### PR TITLE
Support for multiple hosts to pbench-register-tool

### PR DESCRIPTION
**Pull request by portante**
Now the `--remote` parameter has been renamed to `--remotes` to denote that it now accepts one or more remotes on which to register tools, keeping the old `-r|--remote` syntax for compatibility.

Since labels are specified per host, we also add support for a new `--labels` argument, checking to ensure that the number of labels matches the number of remotes specified.

The user can also create a file containing a list of `<hostname>,<label>` pairs (leading `#` comments and blank lines allowed) instead of listing them all on the command line by just specifying `--remotes=@<filename>`.

We also add a new `--no-install` parameter for the user to opt out of running the tool install, in support of environments where tools are already pre-installed (e.g. container images).

The interface changes to add `--remotes`, `--labels`, and `--no-install` applies to `pbench-register-tool-set` as well.

We enhance the unit tests to verify the new behaviors of the interfaces, and also add a test for the `--group` parameter to `pbench-register-tool-set`.

Fixes issue #1285.

### CODE REVIEW

**Comment by k-rister** [comment]
I do like the idea of a uniform scheme for referencing a file as the usage of @ was a little confusing -- I think I would tend to pick the URL style (file://) because you could potentially extend that to support pulling from remote sources, etc.

**Comment by portante** [comment]
Good feedback, let's see if we can come to an agreement.  For starters, the precedent for using the `@` comes from curl [1].

I'd rather not introduce a new way that we are unlikely to move to, so this seems worth-while to hash out now.

[1]  https://unix.stackexchange.com/questions/144479/what-does-the-at-symbol-before-a-filename-mean-in-a-curl-command.

File: agent/util-scripts/pbench-register-tool [hunk 1]
```diff
@@ -1,184 +1,343 @@
 #!/bin/bash
 # -*- mode: shell-script; indent-tabs-mode: t; sh-basic-offset: 8; sh-indentation: 8; sh-indent-for-case-alt: + -*-
 
-script_path=`dirname $0`
-script_name=`basename $0`
-pbench_bin="`cd ${script_path}/..; /bin/pwd`"
+script_path="$(dirname ${0})"
+script_name="$(basename ${0})"
+pbench_bin="$(realpath -e ${script_path}/..)"
```
[the remark this answers is not in the document]
Comment (portante) [reply agent/util-scripts/pbench-register-tool]: Typically `readlink` just reads the symlink and returns the contents.  The `realpath -e` runs through and resolves all symlinks at all levels until the full absolute path is found.  This helps us detect when configurations a broken that provide symlinks to no-where.

File: agent/util-scripts/pbench-register-tool [hunk 2]
```diff
@@ -1,184 +1,343 @@
...
-local_interfaces="$local_ips $hostname $full_hostname localhost"
-# if for some reason ther user specified "--remote=$a" and $a is a local interface,
-# we must remove it, so we don't ssh to ourselves
-if [ ! -z "$remote" ]; then
-	for local_interface in $local_interfaces; do
-		if [ "$remote" == "$local_interface" ]; then
-			debug_log "The remote host you have provided, $remote, matches a local interface, so we will register this tool locally"
+# At this point all argument processing is complete.
+
+# Determine what our local hostname(s) and IP(s) are.
```
[the remark this answers is not in the document]
Comment (portante) [reply agent/util-scripts/pbench-register-tool]: Yes, seems so, perhaps worth making another commit the refactors that out since this code was already here before this commit.

File: agent/util-scripts/pbench-register-tool (hunk 1 above)
[the remark this answers is not in the document]
Comment (k-rister) [reply agent/util-scripts/pbench-register-tool]: Hmm, that is exactly what readlink -e does as well.  Interesting...

[the remark this answers is not in the document]
Comment (portante) [reply agent/util-scripts/pbench-register-tool]: https://unix.stackexchange.com/questions/136494/whats-the-difference-between-realpath-and-readlink-f

**Comment by portante** [comment]
Unit tests fail because of tool listing is not sorted and the `find` command reports differently on different systems; will fix shortly.

**Comment by portante** [comment]
Added the "DO NOT MERGE!!!" label until we come to an agreement on the interfaces proposed here.

**Comment by k-rister** [comment]
> [1] https://unix.stackexchange.com/questions/144479/what-does-the-at-symbol-before-a-filename-mean-in-a-curl-command.

This seems to imply that the @ tells curl to not modify the contents of the file when uploading when it is already expecting a filename.  I don't think that is very relevant to this particular usage (why would we be modifying the file contents?).  That said, curl and other programs (wget, git, browsers, etc.) do use URL protocol syntax to determine how to access files:

```
ftp://
http://
https://
git://
git+ssh://
rsync://
```

**Comment by portante** [comment]
> This seems to imply that the @ tells curl to not modify the contents of the file when uploading when it is already expecting a filename.

The man page for `curl` [1] seems to say:

```
If you start the data with the letter @, the rest should be a
file name to read the data from, or - if you want curl to read
the data from stdin. Multiple files can also be specified.
Posting data from a file named 'foobar' would thus be done
with -d, --data @foobar. When --data is told to read from a
file like that, carriage returns and newlines will be stripped
out. If you don't want the @ character to have a special
interpretation use --data-raw instead.
```

[1] http://man7.org/linux/man-pages/man1/curl.1.html

**Comment by portante** [comment]
The URL style scheme would be fine to use, so something like `--remotes @file:///tmp/remotes.lis` or `--remotes @http://host.example.com/config/remotes` would be powerful.  Not sure we'd support anything other than `file` in this PR, but the scheme would give us the ability to extend easily in the future.

**Comment by atheurer** [comment]
I would prefer the interface which allows for more flexibility later, aka universal resource locator.  We start with file:// and we can expand easily to other types of URLs, even pass directly to curl and let curl handle it all

**Comment by portante** [comment]
Can we use the leading `@` to denote when to process the arguments as a URL versus a comma separated list of hosts?  It makes the behavior in the code clear so that we don't have to check for all the schemes ahead of time.

**Comment by k-rister** [comment]
Wouldn't it just be sufficient to check for `://` in the string?

Are multiple files supported?  Something like:

`--remotes file:///path/to/set1,file:///path/to/set2,file:///path/to/set3`

This would seem really nice for composing groups of systems.

**Comment by portante** [comment]
Wouldn't `--remotes @file:///path/to/set1,file:///path/to/set2,file:///path/to/set3` also work?  It would be umabiguous, leading `@` determines which mode to use, "URL" vs "in-line".

Seems like the decision point is:

- Provide one `--<option>` to handle both in-line data and URL data
- Provide two `--<option>` / --<option-from-URL>` switch, one to handle in-line data and one to handle URL data

If we want one option to handle both in-line and URL data, then we need a clear and unambiguous way for the users to indicate which form data is going to be provided.  Using a leading `@` is a very simple way to remove ambiguity, has precedent with other commands outside of pbench-agent, and is easy for users to use.

I am not in favor of "sniffing" the contents of the parameter beyond the first character because it is not easy to get right in all situations outside of this one.  I agree it is viable for this situation, but it seems the discussion here is how to consider the pattern for all options going forward.

If we want to use two options, then the two options remove the ambiguity.  I think we already do that with `pbench-uperf` with `--clients-file` vs `--clients`.

**Comment by k-rister** [comment]
> Wouldn't `--remotes @file:///path/to/set1,file:///path/to/set2,file:///path/to/set3` also work? It would be umabiguous, leading `@` determines which mode to use, "URL" vs "in-line".
> 
> Seems like the decision point is:
> 
> * Provide one `--<option>` to handle both in-line data and URL data
> * Provide two `--<option>` / --` switch, one to handle in-line data and one to handle URL data

I think one option is much more preferable from a simplicity point of view.

> 
> If we want one option to handle both in-line and URL data, then we need a clear and unambiguous way for the users to indicate which form data is going to be provided. Using a leading `@` is a very simple way to remove ambiguity, has precedent with other commands outside of pbench-agent, and is easy for users to use.

I guess this is where we disagree, I've been using CLI tools for a really, really long time (too long, it hurts to think about it) and I can't ever remember using the `@` for this purpose -- it seems completely foreign and the opposite of unambiguous (I guess that's ambiguous) to me.  The precedent you cite is one relatively obscure option on a single command.  If we weren't having this discussion and I saw that (as I did in your first post) then I would have no idea what it meant.

On the other hand, the `<protocol>://` prefix seems quite clear and very widely used.

> 
> I am not in favor of "sniffing" the contents of the parameter beyond the first character because it is not easy to get right in all situations outside of this one. I agree it is viable for this situation, but it seems the discussion here is how to consider the pattern for all options going forward.

Not sure I see the problem here, a simple regex would do the trick....of course I like Perl and think regex are kewl ;-)

> 
> If we want to use two options, then the two options remove the ambiguity. I think we already do that with `pbench-uperf` with `--clients-file` vs `--clients`.

At the end of the day I don't really feel to strongly about this, but my $0.02 is if you want it to be clear and obvious then I don't think the `@` does that.

**Comment by portante** [comment]
> I think one option is much more preferable from a simplicity point of view.

Great, so one option.  Now we have to figure out how the user indicates to the tool what mode the option is working in: "in-line" vs "URL".

> On the other hand, the <protocol>:// prefix seems quite clear and very widely used.

I don't disagree with this.

> Not sure I see the problem here, a simple regex would do the trick....of course I like Perl and think regex are kewl ;-)

I do disagree with this.  While it will work in this case, if we roll this out to all the cases, each way to indicate URL vs in-line will be different.  By adopting a leading character like `@`, all options behave the same way, and it is very simple to implement (see the existing code which does this is `bash` already).

So, to summarize:

Proposal 1: Use a leading `@` character, the presence of which indicates "what follows is one or more URLs", the absence of which indicates "what follows in in-line data"

Proposal 2: Sniff the contents of the option to see if it contains the pattern, "`[a-z]+://`", the presence of which indicates the option contains "one or more URLs", the absence of which indicates the option contains "in-line data"

Can we agree that these are the two proposals and their meanings?  And that both of them are using the URL syntax for file specification?

I am advocating "proposal 1".

**Comment by k-rister** [comment]
> 
> Great, so one option. Now we have to figure out how the user indicates to the tool what mode the option is working in: "in-line" vs "URL".

Maybe we've been talking past each other because I thought the scheme made the user's intent obvious.  Wouldn't the presence (or lack thereof) of the `<protocol>://` prefix answer that question for you?

```
--remotes=list,of,oddly,named,systems
  vs.
--remotes=file:///path/to/system/list/file
  vs.
--remotes=http://fqdn/path/to/system/list/file
```

**Comment by atheurer** [comment]
I am not totally opposed to the "@"; however, IMO it does not seem necessary to me.  I can go either way I guess.

**Comment by portante** [comment]
I agree with both of you that for *this* option one *could* inspect the contents of the option data to determine the intention of the user.

For the general case of all pbench-agent command line options, I am advocating that we adopt using a prefix character to allow the user to explicitly indicate the kind of data it is providing to the interface.

I am not in favor of interfaces "figuring out" the intention of the user.  I would rather see us provide ways for the user to be explicit about their intention without relying on the semantic contents of the option data to do so.

Does that make sense?

**Comment by k-rister** [comment]
The user specifying a `<protocol>://` seems pretty explicit to me, but like I said before I don't feel that strongly about it.  Either works.

**Comment by portante** [comment]
> Either works.

Thanks @k-rister, I am getting back to this work, will rebase, and finalize.

**Comment by portante** [comment]
@ndokos, this work has been rebased.  Can you review?  I'd like to get this merged as soon as possible so that I can work on the rest of the "Tool Meister" effort (#1248).

File: agent/util-scripts/pbench-register-tool [hunk 3]
```diff
@@ -1,184 +1,343 @@
...
-	this_tool_file="$pbench_run/tools-$group/remote@$remote"
-	/bin/rm -f "$this_tool_file"
-	# the specific tool options are stored on the remote host, so there is no need to keep them here
-	# however, we do use this file to store the label for this remote host [if any]
-	if [ -z "$label" ]; then
-		touch "$this_tool_file"
+	this_tool_file="${pbench_run}/tools-${group}/remote@${remote}"
+	/bin/rm -f "${this_tool_file}"
+	# The specific tool options are stored on the remote host, so there is
+	# no need to keep them here however, we do use this file to store the
```
Comment (ndokos) [inline agent/util-scripts/pbench-register-tool]: Misplaced comma: "... keep them here, however we do use..."

File: agent/util-scripts/pbench-register-tool-set [hunk 4]
```diff
@@ -1,119 +1,153 @@
...
-. "$pbench_bin"/base
+. "${pbench_bin}"/base
 
-# The sole purpose of this script is to ensure a tool
-# of your choosing is used during the execution of a benchmark.
-# For a list of performance tools, look at the
-# $pbench_bin/tool-scripts directory.
+# The sole purpose of this script is to ensure a tool set of your choosing is
+# used during the execution of a benchmark on the optional list of remote
+# hostnames specified. For a list of available tool sets, use the --help
```
Comment (ndokos) [inline agent/util-scripts/pbench-register-tool-set]: "... available tools" perhaps? But --help does not seem to give that list.

File: agent/util-scripts/pbench-register-tool-set [hunk 5]
```diff
@@ -1,119 +1,153 @@
...
 	default_interval=3
 fi
 
 function usage() {
-	printf "usage:\n"
-	printf "$script_name --toolset=<tool-set> [--group=<group-name>] [--remote=<remote-host>] [--label=<label>] [--interval=<interval] -- [all tool specific options here]\n"
+	printf -- "usage:\n"
+	printf -- "${script_name} --toolset=<tool-set> [--group=<group-name>] [--interval=<interval>] [--no-install] [--remotes=<remote-host>[,<remote-host>]] [--labels=<label>[,<label>]]\n"
+	printf -- "\tWhere the list of labels must match the list of remotes.\n"
+	printf -- "\tLike pbench-register-tool, one can specify a single remote\n"
```
Comment (ndokos) [inline agent/util-scripts/pbench-register-tool-set]: I'm willing to live with "a single remote" but I thiink it would be better to enumerate the possibilities as I suggested for `pbench-register-tool`: a single remote, a comma-separated list of remotes, or an at sign with a filename.

File: agent/util-scripts/pbench-register-tool-set [hunk 6]
```diff
@@ -1,119 +1,153 @@
...
+	printf -- "ERROR: failed to fetch tool set, ${toolset}-tool-set, from the pbench-agent configuration file\n" >&2
+	exit 1
+fi
+if [[ -z ${the_tool_set} ]]; then
+	printf -- "ERROR: empty tool set, ${toolset}-tool-set, fetched from the pbench-agent configuration file\n" >&2
+	exit 1
+fi
+
+if [[ -n "${labels_arg}" ]]; then
+	pbench-register-tool --name=perf --test-labels ${remotes_arg}${labels_arg}
```
Comment (ndokos) [inline agent/util-scripts/pbench-register-tool-set]: What's `--test-labels`?

File: agent/util-scripts/pbench-register-tool-set (hunk 4 above)
Comment (portante) [reply agent/util-scripts/pbench-register-tool-set]: We don't register tools with `pbench-register-tool-set`, so I'll just get the available list with `--help`.

File: agent/util-scripts/pbench-register-tool-set (hunk 6 above)
Comment (portante) [reply agent/util-scripts/pbench-register-tool-set]: This is just a way for `pbench-register-tool-set` to check the labels are provided correctly without re-implementing the logic already contained in `pbench-register-tool`.

File: agent/util-scripts/pbench-register-tool [hunk 7]
```diff
@@ -1,184 +1,343 @@
...
+	printf -- "${script_name}: you specified an invalid option\n\n" >&2
+	usage >&2
 	exit 1
 fi
-# here we process only the arguments before the "--"
-# these are argument for pbench-register-tools
-# all agruments after "--" are for the tool itself
-eval set -- "$opts";
+
+# Process only the arguments before the "--": these are argument for
```
[the remark this answers is not in the document]
Comment (portante) [reply agent/util-scripts/pbench-register-tool]: Thanks, fixed.

File: agent/util-scripts/pbench-register-tool [hunk 8]
```diff
@@ -1,184 +1,343 @@
...
+# pbench-register-tool all agruments after "--" are for the tool itself.
```
[the remark this answers is not in the document]
Comment (portante) [reply agent/util-scripts/pbench-register-tool]: Thanks, fixed.

File: agent/util-scripts/pbench-register-tool [hunk 9]
```diff
@@ -1,184 +1,343 @@
...
+# the ${pbench_bin}/tool-scripts directory.
 
 # Defaults
-remote=""
-group=default
+remotes_arg="${full_hostname}"
+group="default"
 options=""
+do_install=1
+_to_test_labels=0
```
[the remark this answers is not in the document]
Comment (portante) [reply agent/util-scripts/pbench-register-tool]: Thanks, fixed.

File: agent/util-scripts/pbench-register-tool (hunk 3 above)
Comment (portante) [reply agent/util-scripts/pbench-register-tool]: Thanks, fixed.

File: agent/util-scripts/pbench-register-tool [hunk 10]
```diff
@@ -1,184 +1,343 @@
...
-		break
+if [[ ${remotes_arg::1} == "@" ]]; then
+	# We are dealing with a file containing the list of remotes.
+
+	if [[ -n ${labels_arg} ]]; then
+		error_log "--labels=${labels_arg} not allowed with remotes file (${remotes_arg})"
+		usage >&2
+		exit 1
+	fi
+	remotes_file=${remotes_arg##*@}
```
[the remark this answers is not in the document]
Comment (portante) [reply agent/util-scripts/pbench-register-tool]: Thanks, fixed.

File: agent/util-scripts/pbench-register-tool [hunk 11]
```diff
@@ -1,184 +1,343 @@
...
+do_install=1
+_to_test_labels=0
 
 function usage() {
-	printf "usage:\n"
-	printf "$script_name --name=<tool-name> [--group=<group-name>] [--remote=<remote-host>] -- [all tool specific options here]\n"
-	printf "for a list of tool specific options, run $pbench_bin/tool-scripts/<tool-name> --help\n"
+	printf -- "usage:\n"
+	printf -- "${script_name} --name=<tool-name> [--group=<group-name>] [--no-install] [--remotes=<remote-host>[,<remote-host>]] [--labels=<label>[,<label>]] -- [all tool specific options here]\n"
+	printf -- "Where the list of labels must match the list of remotes.\n"
```
[the remark this answers is not in the document]
Comment (portante) [reply agent/util-scripts/pbench-register-tool]: Fixed.

File: agent/util-scripts/pbench-register-tool [hunk 12]
```diff
@@ -1,184 +1,343 @@
...
+	printf -- "\tOne can specify a single remote host with a leading\n"
```
[the remark this answers is not in the document]
Comment (portante) [reply agent/util-scripts/pbench-register-tool]: Fixed.

File: agent/util-scripts/pbench-register-tool-set (hunk 5 above)
Comment (portante) [reply agent/util-scripts/pbench-register-tool-set]: Fixed.

File: agent/util-scripts/pbench-register-tool-set (hunk 4 above)
Comment (portante) [reply agent/util-scripts/pbench-register-tool-set]: Thanks; updated with output that now displays the available tool sets.
